microk8s-resources/wrappers/microk8s-refresh-certs.wrapper

```diff
@@ -73,6 +80,43 @@ function revert {
   $SNAP/microk8s-start.wrapper
 }
 
+function refresh_ca {
+  echo "Backing up CA cert and key under $BACKUP"
+  run_with_sudo mkdir -p $SNAP_DATA/var/log/ca-backup/
+  if [ -e ${SNAP_DATA}/certs/ca.crt ]
+  then
+      run_with_sudo cp ${SNAP_DATA}/certs/ca.crt $BACKUP
+  fi
+  if [ -e ${SNAP_DATA}/certs/ca.key ]
+  then
+      run_with_sudo cp ${SNAP_DATA}/certs/ca.key $BACKUP
+  fi
+
+  echo "Creating new CA certificate and key"
+  run_with_sudo rm -f ${SNAP_DATA}/certs/ca.crt
+  run_with_sudo rm -f ${SNAP_DATA}/certs/ca.key
+
+  if [ -e ${CAPATH}/ca.crt ]
+  then
+      run_with_sudo cp ${CAPATH}/ca.crt ${SNAP_DATA}/certs/ca.crt
+  else
+      echo "Could not find ca.crt in $CAPATH, aborting cert update"
+      run_with_sudo cp ${BACKUP}/ca.crt ${SNAP_DATA}/certs/ca.crt
+      exit 1
+  fi
+
+  if [ -e ${CAPATH}/ca.key ]
+  then
+      run_with_sudo cp ${CAPATH}/ca.key ${SNAP_DATA}/certs/ca.key
+  else
+      echo "Could not find ca.key in $CAPATH, aborting cert update"
+      run_with_sudo cp ${BACKUP}/ca.crt ${SNAP_DATA}/certs/ca.crt
+      run_with_sudo cp ${BACKUP}/ca.key ${SNAP_DATA}/certs/ca.key
+      exit 1
+  fi
+
+  update_configs
```

Comment:
With this update_configs we skip the call to `produce_certs` [1]. However the `produce_certs` is using the CA to issue other certificates, see for example [2] where we sign the server certificate with the CA. If we update the CA the rest of the certificates need to be re-issued.

My suggestion is to add the CA files as an optional argument to `produce_certs`. If no argument is provided then `produce_certs`  will have the original behavior. If we provide the CA to the `produce_certs` then we will re-issue the certificates following the execution path of a `refresh` but instead of creating the CA files in `produce_certs`, use the CA files provided. 

[1] https://github.com/ubuntu/microk8s/pull/1423/files#diff-0805575b10502a2e22b62c908aeedf1fL88
[2] https://github.com/ubuntu/microk8s/blob/master/microk8s-resources/actions/common/utils.sh#L369

Reply:
Ok will investigate and change as requested.

Reply:
I have added a line to `refersh_ca` that re-signs the server certificate by calling `gen_server_cert`. This seems to be working well, and fixes the issue of having to restart host system (Thank you). It needs further testing and some investigation on what type of validation and verification of CA cert and key must be done before ingesting them.